widget/edit.controller.js

```diff
@@ -1,77 +1,211 @@
+    $scope.config.entityTrackable = true;
+    $scope.config.layers = $scope.config.layers || [];
+    if ($scope.config.layers.length === 0) {
+      $scope.config.layers.push({
+        label: "",
+        resource: "",
+        sourceNodesField: "",
+        targetNodeField: "",
+        targetNodeSubField: "",
+        targetNodeType: "",
+        targetNodeSubType: ""
+      });
+    }
+
+    if ($scope.config.customModule) {
+      $scope.loadAttributesForCustomModule();
+    }
+
+    function loadAttributesForCustomModule() {
+      $scope.fields = [];
+      $scope.fieldsArray = [];
+      $scope.objectFields = [];
+      var entity = new Entity($scope.config.customModule);
+      entity.loadFields().then(function () {
+        for (var key in entity.fields) {
+          //filtering out JSON fields 
+          if (entity.fields[key].type === "object") {
+            $scope.objectFields.push(entity.fields[key]);
+          }
+        }
+        $scope.fields = entity.getFormFields();
+        angular.extend($scope.fields, entity.getRelationshipFields());
+        $scope.fieldsArray = entity.getFormFieldsArray();
+      });
+    }
+
+    $scope.$watch('config.customModule', function (oldValue, newValue) {
+      if ($scope.config.customModule && oldValue !== newValue) {
+        if ($scope.config.query && $scope.config.query.filters) {
+          delete $scope.config.query.filters;
+        }
+        delete $scope.config.customModuleField;
+        $scope.loadAttributesForCustomModule();
       }
-    };
+    });
+
+    function onChangeModuleType() {
+      changeAttribute();
+      delete $scope.config.query;
+      delete $scope.config.customModuleField;
+      delete $scope.config.customModule;
+    }
 
     function loadModules() {
       appModulesService.load(true).then(function (modules) {
         $scope.modules = modules;
+        $scope.fetchAttributes();
       });
-      if ($scope.config.resource) {
-        $scope.fetchAttributes('target');
-      }
-      if ($scope.config.relation) {
-        $scope.fetchAttributes('relation');
-      }
+      initializeSingleModule();
+    }
+
+    //to load json obect modules
+    function initializeSingleModule() {
+      appModulesService.load(true).then(function (modules) {
+        $scope.modules = modules;
+
+        //Create a list of modules with atleast one JSON field
+        modules.forEach((module) => {
+          var moduleMetaData = modelMetadatasService.getMetadataByModuleType(module.type);
+          for (let fieldIndex = 0; fieldIndex < moduleMetaData.attributes.length; fieldIndex++) {
+            //Check If JSON field is present in the module
+            if (moduleMetaData.attributes[fieldIndex].type === "object") {
+              $scope.jsonObjModuleList.push(module);
+              break;
+            }
+          }
+        })
+      });
+    }
+
+
+    function changeAttribute() {
+      $scope.config.layers = [];
+      $scope.config.layers.push({
```

Comment:
Same object is pushed on line 33. We can remove this redundant code.

Reply:
created a common function to clear values